File: src/VisualBasic/CodeCracker/Refactoring/AllowMembersOrderingCodeFixProvider.StyleCop.vb
Imports System.Composition
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.CodeFixes
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Refactoring

    <ExportCodeFixProvider("StyleCopAllowMembersOrderingCodeFixProvider", LanguageNames.VisualBasic)>
    <[Shared]>
    Public Class StyleCopAllowMembersOrderingCodeFixProvider
        Inherits BaseAllowMembersOrderingCodeFixProvider

        Public Sub New()
            MyBase.New("Order {0}'s members following StyleCop patterns")
        End Sub

        Protected Overrides Function GetMemberDeclarationComparer(document As Document, cancellationToken As CancellationToken) As IComparer(Of DeclarationStatementSyntax)
            Return New StyleCopMembersComparer
        End Function

        Public Class StyleCopMembersComparer
            Implements IComparer(Of DeclarationStatementSyntax)

            ReadOnly typeRank As New Dictionary(Of Type, Integer) From
                {
                    {GetType(FieldDeclarationSyntax), 1},
                    {GetType(ConstructorBlockSyntax), 2},
                    {GetType(DelegateStatementSyntax), 4},
                    {GetType(EventStatementSyntax), 5},
                    {GetType(EventBlockSyntax), 6},
                    {GetType(EnumBlockSyntax), 7},
                    {GetType(InterfaceBlockSyntax), 8},
                    {GetType(PropertyBlockSyntax), 9},
                    {GetType(PropertyStatementSyntax), 9}, ' For VB Auto Properties
                    {GetType(OperatorBlockSyntax), 11},
                    {GetType(MethodBlockSyntax), 12},
                    {GetType(StructureBlockSyntax), 13},
                    {GetType(ClassBlockSyntax), 14}
                }

            ReadOnly specialModifierRank As New Dictionary(Of SyntaxKind, Integer) From
                {
                    {SyntaxKind.ConstKeyword, 1},
                    {SyntaxKind.SharedKeyword, 2}
                }

            ReadOnly accessLevelRank As New Dictionary(Of SyntaxKind, Integer) From
                {
                    {SyntaxKind.PublicKeyword, -4},
                    {SyntaxKind.FriendKeyword, -2},
                    {SyntaxKind.ProtectedKeyword, 1},
                    {SyntaxKind.PrivateKeyword, 2}
                }

            Public Function Compare(x As DeclarationStatementSyntax, y As DeclarationStatementSyntax) As Integer Implements IComparer(Of DeclarationStatementSyntax).Compare
                If x Is Nothing AndAlso y Is Nothing Then Return 0
                If x Is Nothing Then Return 1
                If y Is Nothing Then Return -1
                If x.Equals(y) Then Return 0

                Dim comparedPoints = GetRankPoints(x).CompareTo(GetRankPoints(y))
                    If comparedPoints <> 0 Then Return comparedPoints

                Dim xModifiers = GetModifiers(x)
                Dim yModifiers = GetModifiers(y)
                comparedPoints = GetAccessLevelPoints(xModifiers).CompareTo(GetAccessLevelPoints(yModifiers))
                If comparedPoints <> 0 Then Return comparedPoints

                comparedPoints = GetSpecialModifierPoints(xModifiers).CompareTo(GetSpecialModifierPoints(yModifiers))
                If comparedPoints <> 0 Then Return comparedPoints

                Return GetName(x).CompareTo(GetName(y))
            End Function

            Private Function GetAccessLevelPoints(tokens As SyntaxTokenList) As Integer
                Return sumrankPoints(tokens, accessLevelRank, accessLevelRank(SyntaxKind.PrivateKeyword))
            End Function

            Private Function GetModifiers(node As DeclarationStatementSyntax) As SyntaxTokenList
                If TypeOf node Is MethodBlockBaseSyntax Then
                    Return DirectCast(node, MethodBlockBaseSyntax).Begin.Modifiers
                End If
                If TypeOf node Is FieldDeclarationSyntax Then
                    Return DirectCast(node, FieldDeclarationSyntax).Modifiers
                End If
                If TypeOf node Is DelegateStatementSyntax Then
                    Return DirectCast(node, DelegateStatementSyntax).Modifiers
                End If
                If TypeOf node Is PropertyStatementSyntax Then
                    Return DirectCast(node, PropertyStatementSyntax).Modifiers
                End If
                If TypeOf node Is PropertyBlockSyntax Then
                    Return DirectCast(node, PropertyBlockSyntax).PropertyStatement.Modifiers
                End If
                Return New SyntaxTokenList()
            End Function

            Private Function GetRankPoints(node As DeclarationStatementSyntax) As Integer
                Dim points = 0
                If Not typeRank.TryGetValue(node.GetType(), points) Then
                    Return 0
                End If
                Return points
            End Function

            Private Function GetSpecialModifierPoints(tokens As SyntaxTokenList) As Integer
                Return SumRankPoints(tokens, specialModifierRank, 100)
            End Function

            Private Function SumRankPoints(tokens As SyntaxTokenList, rank As Dictionary(Of SyntaxKind, Integer), defaultSumValue As Integer) As Integer
                Dim points = tokens.Sum(Function(t) If(rank.ContainsKey(t.VBKind), rank(t.VBKind), 0))

                Return If(points = 0, defaultSumValue, points)
            End Function

            Private Function GetName(node As SyntaxNode) As String
                If TypeOf node Is FieldDeclarationSyntax Then
                    Return GetFieldName(DirectCast(node, FieldDeclarationSyntax).Declarators)
                End If
                If TypeOf node Is PropertyStatementSyntax Then
                    Return DirectCast(node, PropertyStatementSyntax).Identifier.Text
                End If
                If TypeOf node Is PropertyBlockSyntax Then
                    Return GetName(DirectCast(node, PropertyBlockSyntax).PropertyStatement)
                End If
                If TypeOf node Is MethodBlockSyntax Then
                    Return DirectCast(node, MethodBlockSyntax).Begin.Identifier.Text
                End If
                If TypeOf node Is SubNewStatementSyntax Then
                    Return "New"
                End If
                If TypeOf node Is EnumBlockSyntax Then
                    Return DirectCast(node, EnumBlockSyntax).EnumStatement.Identifier.Text
                End If
                If TypeOf node Is InterfaceBlockSyntax Then
                    Return DirectCast(node, InterfaceBlockSyntax).Begin.Identifier.Text
                End If
                If TypeOf node Is EventStatementSyntax Then
                    Return DirectCast(node, EventStatementSyntax).Identifier.Text
                End If
                If TypeOf node Is EventBlockSyntax Then
                    Return GetName(DirectCast(node, EventBlockSyntax).EventStatement)
                End If
                If TypeOf node Is OperatorBlockSyntax Then
                    Return DirectCast(node, OperatorBlockSyntax).Begin.OperatorToken.Text
                End If
                If TypeOf node Is DelegateStatementSyntax Then
                    Return DirectCast(node, DelegateStatementSyntax).Identifier.Text
                End If
                If TypeOf node Is ClassBlockSyntax Then
                    Return DirectCast(node, ClassBlockSyntax).Begin.Identifier.Text
                End If
                If TypeOf node Is StructureBlockSyntax Then
                    Return DirectCast(node, StructureBlockSyntax).Begin.Identifier.Text
                End If
                Return ""
            End Function
            Custom Event foo As Action
                AddHandler(value As Action)

                End AddHandler
                RemoveHandler(value As Action)

                End RemoveHandler
                RaiseEvent()

                End RaiseEvent
            End Event


            Private Function GetFieldName(declarations As SeparatedSyntaxList(Of VariableDeclaratorSyntax)) As String
                Dim names = From declaration In declarations
                            From name In declaration.Names
                            Select name.Identifier.Text
                Return String.Join("", names)
            End Function
        End Class
    End Class
End Namespace

File: src/CSharp/CodeCracker/Refactoring/AllowMembersOrderingCodeFixProvider.StyleCop.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System;
using System.Collections.Generic;
using System.Composition;
using System.Linq;
using System.Text;
using System.Threading;

namespace CodeCracker.CSharp.Refactoring
{
    [ExportCodeFixProvider("StyleCopAllowMembersOrderingCodeFixProvider", LanguageNames.CSharp), Shared]
    public class StyleCopAllowMembersOrderingCodeFixProvider : BaseAllowMembersOrderingCodeFixProvider
    {
        public StyleCopAllowMembersOrderingCodeFixProvider() :
            base("Order {0}'s members following StyleCop patterns") { }

        protected override IComparer<MemberDeclarationSyntax> GetMemberDeclarationComparer(Document d, CancellationToken c) =>
            new StyleCopMembersComparer();

        public class StyleCopMembersComparer : IComparer<MemberDeclarationSyntax>
        {
            readonly Dictionary<Type, int> typeRank = new Dictionary<Type, int>
            {
                { typeof(FieldDeclarationSyntax),       1 },
                { typeof(ConstructorDeclarationSyntax), 2 },
                { typeof(DestructorDeclarationSyntax),  3 },
                { typeof(DelegateDeclarationSyntax),    4 },
                { typeof(EventFieldDeclarationSyntax),  5 },
                { typeof(EventDeclarationSyntax),       6 },
                { typeof(EnumDeclarationSyntax),        7 },
                { typeof(InterfaceDeclarationSyntax),   8 },
                { typeof(PropertyDeclarationSyntax),    9 },
                { typeof(IndexerDeclarationSyntax),     10 },
                { typeof(OperatorDeclarationSyntax),    11 },
                { typeof(MethodDeclarationSyntax),      12 },
                { typeof(StructDeclarationSyntax),      13 },
                { typeof(ClassDeclarationSyntax),       14 },
            };

            private readonly Dictionary<SyntaxKind, int> specialModifierRank = new Dictionary<SyntaxKind, int>
            {
                { SyntaxKind.ConstKeyword,   1 },
                { SyntaxKind.StaticKeyword,  2 },
            };

            private readonly Dictionary<SyntaxKind, int> accessLevelRank = new Dictionary<SyntaxKind, int>
            {
                { SyntaxKind.PublicKeyword,     -4 },
                { SyntaxKind.InternalKeyword,   -2 },
                { SyntaxKind.ProtectedKeyword,   1 },
                { SyntaxKind.PrivateKeyword,     2 },
            };

            public int Compare(MemberDeclarationSyntax x, MemberDeclarationSyntax y)
            {
                if (x == null && y == null) return 0;
                if (x == null) return 1;
                if (y == null) return -1;

                var comparedPoints = GetRankPoints(x).CompareTo(GetRankPoints(y));
                if (comparedPoints != 0)
                    return comparedPoints;

                var xModifiers = GetModifiers(x);
                var yModifiers = GetModifiers(y);

                comparedPoints = GetAccessLevelPoints(xModifiers).CompareTo(GetAccessLevelPoints(yModifiers));
                if (comparedPoints != 0)
                    return comparedPoints;

                comparedPoints = GetSpecialModifierPoints(xModifiers).CompareTo(GetSpecialModifierPoints(yModifiers));
                if (comparedPoints != 0)
                    return comparedPoints;

                return comparedPoints != 0 ? comparedPoints : GetName(x).CompareTo(GetName(y));
            }

            private int GetRankPoints(MemberDeclarationSyntax node)
            {
                int points;
                if (!typeRank.TryGetValue(node.GetType(), out points))
                    return 0;
                return points;
            }

            private SyntaxTokenList GetModifiers(MemberDeclarationSyntax node)
            {
                var nodeType = node.GetType();

                if (node is BaseMethodDeclarationSyntax)
                    return (node as BaseMethodDeclarationSyntax).Modifiers;

                if (node is BaseFieldDeclarationSyntax)
                    return (node as BaseFieldDeclarationSyntax).Modifiers;

                if (node is DelegateDeclarationSyntax)
                    return (node as DelegateDeclarationSyntax).Modifiers;

                if (node is BasePropertyDeclarationSyntax)
                    return (node as BasePropertyDeclarationSyntax).Modifiers;

                return default(SyntaxTokenList);
            }

            private int GetSpecialModifierPoints(SyntaxTokenList tokenList) => SumRankPoints(tokenList, specialModifierRank, 100);

            private int GetAccessLevelPoints(SyntaxTokenList tokenList) => SumRankPoints(tokenList, accessLevelRank, accessLevelRank[SyntaxKind.PrivateKeyword]);

            private int SumRankPoints(SyntaxTokenList tokenList, Dictionary<SyntaxKind, int> rank, int defaultSumValue)
            {
                var points = tokenList
                        .Select(s => s.Kind())
                        .Sum(tokenKind => rank.ContainsKey(tokenKind) ? rank[tokenKind] : 0);
                return points == 0 ? defaultSumValue : points;
            }

            private string GetName(SyntaxNode node)
            {
                if (node is FieldDeclarationSyntax)
                    return GetName((node as FieldDeclarationSyntax).Declaration);

                if (node is EventFieldDeclarationSyntax)
                    return GetName((node as EventFieldDeclarationSyntax).Declaration);

                if (node is OperatorDeclarationSyntax)
                    return (node as OperatorDeclarationSyntax).OperatorToken.Text;

                if (node is IndexerDeclarationSyntax)
                    return "this";

                if (node is PropertyDeclarationSyntax)
                    return (node as PropertyDeclarationSyntax).Identifier.Text;

                if (node is MethodDeclarationSyntax)
                    return (node as MethodDeclarationSyntax).Identifier.Text;

                if (node is ConstructorDeclarationSyntax)
                    return (node as ConstructorDeclarationSyntax).Identifier.Text;

                if (node is DestructorDeclarationSyntax)
                    return (node as DestructorDeclarationSyntax).Identifier.Text;

                if (node is DelegateDeclarationSyntax)
                    return (node as DelegateDeclarationSyntax).Identifier.Text;

                if (node is EventDeclarationSyntax)
                    return (node as EventDeclarationSyntax).Identifier.Text;

                if (node is BaseTypeDeclarationSyntax)
                    return (node as BaseTypeDeclarationSyntax).Identifier.Text;

                return "";
            }

            private string GetName(VariableDeclarationSyntax declaration)
            {
                var str = new StringBuilder();
                declaration.Variables.Aggregate(str, (accumulate, seed) => accumulate.Append(seed.Identifier.Text));
                return str.ToString();
            }
        }
    }
}